src/cdk/datetime/datepicker.spec.ts

```diff
@@ -0,0 +1,646 @@
+        expect(testComponent.datepicker._selected).toEqual(selected);
+      }));
+
+      it('should update model when date is selected', fakeAsync(() => {
+        expect(testComponent.selected).toBeNull();
+        expect(testComponent.datepickerInput.value).toBeNull();
+
+        let selected = new Date(2017, JAN, 1);
+        testComponent.datepicker._select(selected);
+        fixture.detectChanges();
+        flush();
+        fixture.detectChanges();
+
+        expect(testComponent.selected).toEqual(selected);
+        expect(testComponent.datepickerInput.value).toEqual(selected);
+      }));
+
+      it('should mark input dirty after input event', () => {
+        let inputEl = fixture.debugElement.query(By.css('input')).nativeElement;
+
+        expect(inputEl.classList).toContain('ng-pristine');
+
+        inputEl.value = '2001-01-01';
+        dispatchFakeEvent(inputEl, 'input');
+        fixture.detectChanges();
+
+        expect(inputEl.classList).toContain('ng-dirty');
+      });
+
+      it('should mark input dirty after date selected', fakeAsync(() => {
+        let inputEl = fixture.debugElement.query(By.css('input')).nativeElement;
+
+        expect(inputEl.classList).toContain('ng-pristine');
+
+        testComponent.datepicker._select(new Date(2017, JAN, 1));
+        fixture.detectChanges();
+        flush();
+        fixture.detectChanges();
+
+        expect(inputEl.classList).toContain('ng-dirty');
+      }));
+
+      it('should not mark dirty after model change', fakeAsync(() => {
+        let inputEl = fixture.debugElement.query(By.css('input')).nativeElement;
+
+        expect(inputEl.classList).toContain('ng-pristine');
+
+        testComponent.selected = new Date(2017, JAN, 1);
+        fixture.detectChanges();
+        flush();
+        fixture.detectChanges();
+
+        expect(inputEl.classList).toContain('ng-pristine');
+      }));
+
+      it('should mark input touched on blur', () => {
+        let inputEl = fixture.debugElement.query(By.css('input')).nativeElement;
+
+        expect(inputEl.classList).toContain('ng-untouched');
+
+        dispatchFakeEvent(inputEl, 'focus');
+        fixture.detectChanges();
+
+        expect(inputEl.classList).toContain('ng-untouched');
+
+        dispatchFakeEvent(inputEl, 'blur');
+        fixture.detectChanges();
+
+        expect(inputEl.classList).toContain('ng-touched');
+      });
+
+      it('should not reformat invalid dates on blur', () => {
+        const inputEl = fixture.debugElement.query(By.css('input')).nativeElement;
+
+        inputEl.value = 'very-valid-date';
+        dispatchFakeEvent(inputEl, 'input');
+        fixture.detectChanges();
+
+        dispatchFakeEvent(inputEl, 'blur');
+        fixture.detectChanges();
+
+        expect(inputEl.value).toBe('very-valid-date');
+      });
+
+      it('should mark input touched on calendar selection', fakeAsync(() => {
+        let inputEl = fixture.debugElement.query(By.css('input')).nativeElement;
```

Comment:
There are a few of these `let inputEl` in some of the tests. Consider switching them to a `const` since they aren't being reassigned.